src/sdk/pynni/nni/bohb_advisor/config_generator.py

```diff
@@ -0,0 +1,326 @@
+                        # so it shouldn't be terrible.
+                        if np.isfinite(l(vector)):
+                            best_vector = vector
+                            break
+
+                    if val < best:
+                        best = val
+                        best_vector = vector
+
+                if best_vector is None:
+                    logger.debug("Sampling based optimization with %i samples failed -> using random configuration"%self.num_samples)
+                    sample = self.configspace.sample_configuration().get_dictionary()
+                    info_dict['model_based_pick'] = False
+                else:
+                    logger.debug('best_vector: {}, {}, {}, {}'.format(best_vector, best, l(best_vector), g(best_vector)))
+                    for i, hp_value in enumerate(best_vector):
+                        if isinstance(
+                            self.configspace.get_hyperparameter(
+                                self.configspace.get_hyperparameter_by_idx(i)
+                            ),
+                            ConfigSpace.hyperparameters.CategoricalHyperparameter
+                        ):
+                            best_vector[i] = int(np.rint(best_vector[i]))
+                    sample = ConfigSpace.Configuration(self.configspace, vector=best_vector).get_dictionary()
+
+                    try:
+                        sample = ConfigSpace.util.deactivate_inactive_hyperparameters(
+                            configuration_space=self.configspace,
+                            configuration=sample)
+                        info_dict['model_based_pick'] = True
+
+                    except Exception as e:
+                        logger.warning(("="*50 + "\n")*3 +\
+                                "Error converting configuration:\n%s"%sample+\
+                                "\n here is a traceback:" +\
+                                traceback.format_exc())
+                        raise(e)
+            except:
+                logger.warning("Sampling based optimization with %i samples failed\n %s \n \
+                Using random configuration"%(self.num_samples, traceback.format_exc()))
+                sample = self.configspace.sample_configuration()
+                info_dict['model_based_pick'] = False
+
+        try:
+            sample = ConfigSpace.util.deactivate_inactive_hyperparameters(
+                configuration_space=self.configspace,
+                configuration=sample.get_dictionary()
+            ).get_dictionary()
+        except Exception as e:
+            logger.warning("Error (%s) converting configuration: %s -> "
+                        "using random configuration!", e, sample)
+            sample = self.configspace.sample_configuration().get_dictionary()
+        logger.debug('done sampling a new configuration.')
+        sample['STEPS'] = budget
+        return sample
+
+    def impute_conditional_data(self, array):
```

Comment:
impute? confuse the naming...

Reply:
This function is used to pick the good point and bad point for each KDE training.
In statistics, impute is the process of replacing missing data with substituted values.
Reference: https://en.wikipedia.org/wiki/Imputation_(statistics)